### ISSUE TITLE
[BUG]: shardformer ShardConfig里的开关不起作用

### ISSUE THREAD

**Issue by wangbluo** [hpcaitech/ColossalAI#4474]
### 🐛 Describe the bug

用了example里给的脚本，将模型替换成7b的llama2，试了shardconfig里的各个优化开关，跑出来的迭代时间都没有发生变化。
脚本：
shardformer_benchmark.py:
`def train(args):
    colossalai.launch_from_torch(config={}, seed=42)
    coordinator = DistCoordinator()

    if args.model == "llama":
        #cfg = BertConfig.from_pretrained(args.pretrain, num_labels=data_builder.num_labels)
        model = LlamaForCausalLM.from_pretrained(args.pretrain)

    model.to(torch.cuda.current_device())

    # if multiple GPUs, shard the model
    if dist.get_world_size() > 1:
        shard_config = ShardConfig(enable_all_optimization=args.all_optimization)
        shard_former = ShardFormer(shard_config=shard_config)
        model = shard_former.optimize(model)

    optim = Adam(model.parameters(), lr=args.lr)
    
    fit(model, optim, args.max_epochs, args.accumulation_steps, args.batch_size, coordinator)
    
    def fit(model: nn.Module, optimizer: Optimizer, max_epochs, accumulation_steps, batch_size,
        coordinator):
    logger = get_dist_logger()
    BATCH_SIZE = args.batch_size
    SEQ_LENGTH = 1024
    VOCAB_SIZE = 32000
    NUM_STEPS = 1000
    total_loss = 0
    global_numel = sum([p.numel() for p in model.parameters()])
    get_tflops_func = partial(get_tflops, global_numel, BATCH_SIZE, SEQ_LENGTH)
    model.train()
    for n in range(1000):
        input_ids, attn_mask = get_data(BATCH_SIZE, SEQ_LENGTH, VOCAB_SIZE)
        input_ids.to('cuda')
        attn_mask.to('cuda')
        model.to('cuda')
        start = time()
        outputs = model(input_ids, attn_mask,None,None,None,input_ids)
        loss = outputs['loss']
        #loss = loss / accumulation_steps
        loss.backward()
        optimizer.step()
        torch.cuda.synchronize()
        step_time = time() - start
        logger.info(
            f'[{n+1}/{NUM_STEPS}] Loss:{loss.item():.3f}, Step time: {step_time:.3f}s, TFLOPS: {get_tflops_func(step_time):.3f}',
            ranks=[0])
    torch.cuda.synchronize()`
    
    shardformer_config.sh:
`torchrun --standalone --nproc_per_node=8 llama2_7b_shardformer_benchmark.py \
    --model "llama" \
    --pretrain "/data0/pkg/Huggingface/Llama-2-7b-hf" \
    --max_epochs 1 \
    --batch_size 8 \
    --lr 2.4e-5 \
    --all_optimization True \
    --accumulation_steps 8 \
    --warmup_fraction 0.03
`
运行结果：
![K20230818-162631](https://github.com/hpcaitech/ColossalAI/assets/32676639/45d47ef9-1574-40e4-b166-6c8c35457cc9)
![K20230818-162643](https://github.com/hpcaitech/ColossalAI/assets/32676639/0c3120f8-39a0-42db-8426-6d7a29a8fc2f)

### Environment

colossalai version：0.3.1
pytorch version: 1.12
CUDA Version: 11.7

**Comment by flybird11111** [comment]
> ### 🐛 Describe the bug
> 用了example里给的脚本，将模型替换成7b的llama2，试了shardconfig里的各个优化开关，跑出来的迭代时间都没有发生变化。 脚本： shardformer_benchmark.py: `def train(args): colossalai.launch_from_torch(config={}, seed=42) coordinator = DistCoordinator()
> 
> ```
> if args.model == "llama":
>     #cfg = BertConfig.from_pretrained(args.pretrain, num_labels=data_builder.num_labels)
>     model = LlamaForCausalLM.from_pretrained(args.pretrain)
> 
> model.to(torch.cuda.current_device())
> 
> # if multiple GPUs, shard the model
> if dist.get_world_size() > 1:
>     shard_config = ShardConfig(enable_all_optimization=args.all_optimization)
>     shard_former = ShardFormer(shard_config=shard_config)
>     model = shard_former.optimize(model)
> 
> optim = Adam(model.parameters(), lr=args.lr)
> 
> fit(model, optim, args.max_epochs, args.accumulation_steps, args.batch_size, coordinator)
> 
> def fit(model: nn.Module, optimizer: Optimizer, max_epochs, accumulation_steps, batch_size,
>     coordinator):
> logger = get_dist_logger()
> BATCH_SIZE = args.batch_size
> SEQ_LENGTH = 1024
> VOCAB_SIZE = 32000
> NUM_STEPS = 1000
> total_loss = 0
> global_numel = sum([p.numel() for p in model.parameters()])
> get_tflops_func = partial(get_tflops, global_numel, BATCH_SIZE, SEQ_LENGTH)
> model.train()
> for n in range(1000):
>     input_ids, attn_mask = get_data(BATCH_SIZE, SEQ_LENGTH, VOCAB_SIZE)
>     input_ids.to('cuda')
>     attn_mask.to('cuda')
>     model.to('cuda')
>     start = time()
>     outputs = model(input_ids, attn_mask,None,None,None,input_ids)
>     loss = outputs['loss']
>     #loss = loss / accumulation_steps
>     loss.backward()
>     optimizer.step()
>     torch.cuda.synchronize()
>     step_time = time() - start
>     logger.info(
>         f'[{n+1}/{NUM_STEPS}] Loss:{loss.item():.3f}, Step time: {step_time:.3f}s, TFLOPS: {get_tflops_func(step_time):.3f}',
>         ranks=[0])
> torch.cuda.synchronize()`
> 
> shardformer_config.sh:
> ```
> 
> `torchrun --standalone --nproc_per_node=8 llama2_7b_shardformer_benchmark.py \ --model "llama" \ --pretrain "/data0/pkg/Huggingface/Llama-2-7b-hf" \ --max_epochs 1 \ --batch_size 8 \ --lr 2.4e-5 \ --all_optimization True \ --accumulation_steps 8 \ --warmup_fraction 0.03 ` 运行结果： ![K20230818-162631](https://user-images.githubusercontent.com/32676639/261547745-45d47ef9-1574-40e4-b166-6c8c35457cc9.png) ![K20230818-162643](https://user-images.githubusercontent.com/32676639/261547783-0c3120f8-39a0-42db-8426-6d7a29a8fc2f.png)
> 
> ### Environment
> colossalai version：0.3.1 pytorch version: 1.12 CUDA Version: 11.7

https://github.com/hpcaitech/ColossalAI/blob/main/colossalai/shardformer/README.md You can refer to the benchmark section in our documentation. When dealing with longer sequences, the time optimization effects become more pronounced.

**Comment by wangbluo** [comment]
@flybird11111 Thank you for your comment, I will update my code to the latest and try again
